[RenderStyleGen] Generate getter/setter functions that were missed

weinig opened the pull request:
#### 8aa1e5bdbca5cd799671d9336bfc488ed7f8de3e
<pre>
[RenderStyleGen] Generate getter/setter functions that were missed
<a href="https://bugs.webkit.org/show_bug.cgi?id=303398">https://bugs.webkit.org/show_bug.cgi?id=303398</a>

Reviewed by Darin Adler.

While I knew that a few classes of properties still did not have generated
getter/setters, I hadn&apos;t noticed that there were a bunch that were missed
just due to them being implemented in purely in RenderStyle.h. To ensure
that I didn&apos;t miss any more, I added a check to the generator that raises
an exception if the metadata necessary for getter/setter generation is not
available. As an escape hatch, we now exclude the following classes of
properties for that check:
    - has the new &quot;skip-render-style&quot; entry (needed for a few internal properties that don&apos;t have any storage, and for tests)
    - is a logical property
    - is a shorthand property
    - is a cascade alias
    - is a coordinated list value

For all other properties, either a storage path or being marked as having
a custom getter/setter is now required.

Due to some properties getter/setters now being available via RenderStyleInlines.h
and RenderStyleSetters.h, rather than RenderStyle.h itself, a bit of shuffling was
needed to avoid including those files in non-Inlines headers.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/history/CachedFrame.cpp:
* Source/WebCore/layout/formattingContexts/block/tablewrapper/TableWrapperBlockFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineContentAligner.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
* Source/WebCore/layout/layouttree/LayoutBox.cpp:
* Source/WebCore/layout/layouttree/LayoutBox.h:
* Source/WebCore/layout/layouttree/LayoutBoxInlines.h:
* Source/WebCore/rendering/LegacyInlineBox.h:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxInlines.h:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerInlines.h:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderLayerModelObjectInlines.h: Added.
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/303805@main">https://commits.webkit.org/303805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea19ac73dc1d02351c6e78c5b54411d421940360

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44802 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141203 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6007 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102227 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136585 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119804 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83027 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/132987 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4589 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113725 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143851 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5815 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38498 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110610 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4974 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110794 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28095 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116056 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59564 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5867 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34367 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5715 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69325 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5822 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1428,18 +1424,18 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle, WTF::DefaultedOp
     inline void setImageRendering(ImageRendering);
 
-    void setWhiteSpaceCollapse(WhiteSpaceCollapse v) { m_inheritedFlags.whiteSpaceCollapse = static_cast<unsigned>(v); }
+    inline void setWhiteSpaceCollapse(WhiteSpaceCollapse v);
```
```suggestion
    inline void setWhiteSpaceCollapse(WhiteSpaceCollapse);
```

darinadler commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -572,7 +572,7 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle, WTF::DefaultedOp
 
-    constexpr DisplayType display() const { return static_cast<DisplayType>(m_nonInheritedFlags.effectiveDisplay); }
+    inline DisplayType display() const;
     constexpr WritingMode writingMode() const { return m_inheritedFlags.writingMode; }
     bool isLeftToRightDirection() const { return writingMode().isBidiLTR(); } // deprecated, because of confusion between physical inline directions and bidi / line-relative directions
```
Why not also move functions like these into the `Inlines.h` file? How are you drawing the line? (Maybe this should be obvious if I read your change log comment carefully enough.)

weinig replied:
No real reason other than making these changes in stages.

darinadler commented on Source/WebCore/layout/layouttree/LayoutBox.h:
```diff
@@ -127,34 +127,34 @@ class Box : public CanMakeCheckedPtr<Box> {
     bool isInterlinearRubyAnnotationBox() const;
 
     bool isDocumentBox() const { return m_nodeType == NodeType::DocumentElement; }
```
I think it’s worth considering moving even more of these into the `Inlines.h` header; not sure what the pros and cons would be.

weinig replied:
I don't think for these there is any benefit to the number includes, since NodeType is a nested type in LayoutBox, but perhaps it would be less straining on the compiler? I really don't know what the tradeoff for a trivial function like this is in that case.

weinig committed:
[RenderStyleGen] Generate getter/setter functions that were missed
https://bugs.webkit.org/show_bug.cgi?id=303398

Reviewed by Darin Adler.

While I knew that a few classes of properties still did not have generated
getter/setters, I hadn't noticed that there were a bunch that were missed
just due to them being implemented in purely in RenderStyle.h. To ensure
that I didn't miss any more, I added a check to the generator that raises
an exception if the metadata necessary for getter/setter generation is not
available. As an escape hatch, we now exclude the following classes of
properties for that check:
    - has the new "skip-render-style" entry (needed for a few internal properties that don't have any storage, and for tests)
    - is a logical property
    - is a shorthand property
    - is a cascade alias
    - is a coordinated list value

For all other properties, either a storage path or being marked as having
a custom getter/setter is now required.

Due to some properties getter/setters now being available via RenderStyleInlines.h
and RenderStyleSetters.h, rather than RenderStyle.h itself, a bit of shuffling was
needed to avoid including those files in non-Inlines headers.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/history/CachedFrame.cpp:
* Source/WebCore/layout/formattingContexts/block/tablewrapper/TableWrapperBlockFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineContentAligner.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
* Source/WebCore/layout/layouttree/LayoutBox.cpp:
* Source/WebCore/layout/layouttree/LayoutBox.h:
* Source/WebCore/layout/layouttree/LayoutBoxInlines.h:
* Source/WebCore/rendering/LegacyInlineBox.h:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxInlines.h:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerInlines.h:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderLayerModelObjectInlines.h: Added.
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/303805@main

csaavedra commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -572,7 +572,7 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle, WTF::DefaultedOp
     // attribute getter methods
 
-    constexpr DisplayType display() const { return static_cast<DisplayType>(m_nonInheritedFlags.effectiveDisplay); }
+    inline DisplayType display() const;
```
The `display()` method lost `constexpr` here and that causes issues with older GCC versions that are more strict with `constexpr` methods not being allowed to call non-`constexpr` functions, see https://build.webkit.org/#/builders/46/builds/58096

lauromoura replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=303504 with this information about the failure.

dpino replied:
proposed build fix: https://github.com/WebKit/WebKit/pull/54827
